File: python/linked-list/linked_list.py
class Node(object):
    def __init__(self, value, succeeding=None, previous=None):
        self._value = value
        self._prev = previous
        self._next = succeeding


class LinkedList(object):
    def __init__(self):
        self._head = None
        self._end = None
    
    def push(self, value):
        node = Node(value, None, self._end)
        if self._end is not None:
            self._end._next = node
        if self._head is None:
            self._head = node
        self._end = node
    
    def pop(self):
        node = self._end
        self._end = self._end._prev
        if self._end is not None:
            self._end._next = None
        else:
            self._head = None
        return node._value
    
    def unshift(self, value):
        node = Node(value, self._head, None)
        if self._head is not None:
            self._head._prev = node
        if self._end is None:
            self._end = node
        self._head = node

    def shift(self):
        node = self._head
        self._head = self._head._next
        if self._head is not None:
            self._head._prev = None
        else:
            self._end = None
        return node._value
    
    def __iter__(self):
        pointer = self._head
        while pointer:
            yield pointer._value
            pointer = pointer._next
    
    def __len__(self):
        counter = 0
        for _ in self:
            counter += 1
        return counter

File: ruby/linked-list/linked_list.rb
class Deque

  attr_reader :first_node

  def push(element)
    new_node = Node.new(element)
    if @first_node.nil?
      @first_node = new_node
    else
      new_node.previous = @first_node.last
      new_node.previous.next = new_node
    end
  end

  def unshift(element)
    new_node = Node.new(element)
    if @first_node.nil?
      @first_node = new_node
    else
      new_node.next = @first_node
      new_node.next.previous = new_node
      @first_node = new_node
    end
  end


  def pop
    last_node = @first_node.last
    if @first_node.next.nil?
      @first_node = nil
    else
      # break relationships
      last_node.previous.next = nil
      last_node.previous = nil
    end
    last_node.value
  end

  def shift
    first_node = @first_node
    if @first_node.next.nil?
      @first_node = nil
    else
      # break relationships
      @first_node = first_node.next
      first_node.next.previous = nil
      first_node.next = nil
    end
    first_node.value
  end


  class Node
    attr_accessor :value, :next, :previous
    def initialize(value)
      @value = value
    end

    def last
      return self if @next.nil?
      pointer = @next
      pointer = pointer.next until pointer.next.nil?
      pointer
    end

  end
end
